chain/actors/actor_storagemarket.go

```diff
@@ -0,0 +1,541 @@
+package actors
+
+import (
+	"bytes"
+
+	"github.com/filecoin-project/go-amt-ipld"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-hamt-ipld"
+	cbg "github.com/whyrusleeping/cbor-gen"
+
+	"github.com/filecoin-project/go-lotus/build"
+	"github.com/filecoin-project/go-lotus/chain/actors/aerrors"
+	"github.com/filecoin-project/go-lotus/chain/address"
+	"github.com/filecoin-project/go-lotus/chain/types"
+)
+
+type StorageMarketActor struct{}
+
+type smaMethods struct {
+	Constructor                  uint64
+	WithdrawBalance              uint64
+	AddBalance                   uint64
+	CheckLockedBalance           uint64
+	PublishStorageDeals          uint64
+	HandleCronAction             uint64
+	SettleExpiredDeals           uint64
+	ProcessStorageDealsPayment   uint64
+	SlashStorageDealCollateral   uint64
+	GetLastExpirationFromDealIDs uint64
+	ActivateStorageDeals         uint64
+}
+
+var SMAMethods = smaMethods{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11}
+
+func (sma StorageMarketActor) Exports() []interface{} {
+	return []interface{}{
+		2: sma.WithdrawBalance,
+		3: sma.AddBalance,
+		// 4: sma.CheckLockedBalance,
+		5: sma.PublishStorageDeals,
+		// 6: sma.HandleCronAction,
+		// 7: sma.SettleExpiredDeals,
+		// 8: sma.ProcessStorageDealsPayment,
+		// 9: sma.SlashStorageDealCollateral,
+		// 10: sma.GetLastExpirationFromDealIDs,
+		11: sma.ActivateStorageDeals, // TODO: move under PublishStorageDeals after specs team approves
+	}
+}
+
+type StorageParticipantBalance struct {
+	Locked    types.BigInt
+	Available types.BigInt
+}
+
+type StorageMarketState struct {
+	Balances cid.Cid // hamt<addr, StorageParticipantBalance>
+	Deals    cid.Cid // amt<StorageDeal>
+
+	NextDealID uint64 // TODO: amt.LastIndex()
+}
+
+type StorageDealProposal struct {
+	PieceRef           cid.Cid // can this mess anything up? Should this just be cid bytes
```

Comment:
heh, including piece refs like this can make naive chain walkers very sad. i'd push for just cid bytes